src/move/market_metadata/sources/emojicoin_dot_fun_market_metadata.move

```diff
@@ -0,0 +1,624 @@
+        market: address, property: String
+    ): Option<String> acquires Registry {
+        let map =
+            &Registry[@market_metadata].map;
+        if (!map.contains(market)) {
+            return option::none();
+        };
+        let market_properties = map.borrow(market);
+        if (!market_properties.contains_key(&property)) {
+            return option::none();
+        };
+        option::some(*market_properties.borrow(&property))
+    }
+
+    public entry fun add_admin(admin: &signer, new_admin: address) acquires Registry {
+        let admins_ref_mut = &mut borrow_registry_mut_checked(admin).admins;
+        assert!(!admins_ref_mut.contains(&new_admin), E_ALREADY_ADMIN);
+        admins_ref_mut.push_back(new_admin);
+    }
+
+    public entry fun remove_admin(admin: &signer, admin_to_remove: address) acquires Registry {
+        let admins_ref_mut = &mut borrow_registry_mut_checked(admin).admins;
+        let (admin_to_remove_is_admin, admin_to_remove_index) =
+            admins_ref_mut.index_of(&admin_to_remove);
+        assert!(
+            admin_to_remove != @market_metadata,
+            E_ADMIN_TO_REMOVE_IS_MARKET_METADATA_PUBLISHER
+        );
+        assert!(admin_to_remove_is_admin, E_ADMIN_TO_REMOVE_IS_NOT_ADMIN);
+        admins_ref_mut.remove(admin_to_remove_index);
+    }
+
+    public entry fun add_market_properties(
+        admin: &signer,
+        market: address,
+        keys: vector<String>,
+        values: vector<String>
+    ) acquires Registry {
+        assert!(
+            option::is_some(&emojicoin_dot_fun::market_metadata_by_market_address(market)),
+            E_MARKET_NOT_FOUND
+        );
+        assert!(keys.length() == values.length(), E_KEYS_AND_VALUES_NOT_EQUAL_LENGTH);
+        let vault_ref_mut = borrow_registry_mut_checked(admin);
+        let map = &mut vault_ref_mut.map;
+        if (!map.contains(market)) {
+            map.add(market, simple_map::create());
+        };
+        let market_properties = map.borrow_mut(market);
+        while (keys.length() > 0) {
+            market_properties.upsert(keys.pop_back(), values.pop_back());
+        }
+    }
+
+    public entry fun set_market_properties(
+        admin: &signer,
+        market: address,
+        keys: vector<String>,
+        values: vector<String>
+    ) acquires Registry {
+        assert!(
+            option::is_some(&emojicoin_dot_fun::market_metadata_by_market_address(market)),
+            E_MARKET_NOT_FOUND
+        );
+        assert!(keys.length() == values.length(), E_KEYS_AND_VALUES_NOT_EQUAL_LENGTH);
+        let vault_ref_mut = borrow_registry_mut_checked(admin);
+        let map = &mut vault_ref_mut.map;
+        map.upsert(market, simple_map::create());
+        let market_properties = map.borrow_mut(market);
+        while (keys.length() > 0) {
+            market_properties.upsert(keys.pop_back(), values.pop_back());
+        }
+    }
+
+    public entry fun remove_market_properties(
```

Comment:
```suggestion
    /// If the market has an entry, remove all specified keys. If the market has no entries after
    /// removing the keys, remove the market from the metadata registry.
    public entry fun remove_market_properties(
```

Reply:
Addressed in f3362f52f8c192548676036fcdfcc3312dc15162.